File: sol/solution/0001-0100/0025/Solution.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? reverseKGroup(ListNode? head, int k) {
    ListNode? curr = head;
    int count = 0;
    while (curr != null && count < k) {
      curr = curr.next;
      count++;
    }
    if (count == k) {
      final reversedHead = reverseLinkedList(head, k);
      head?.next = reverseKGroup(curr, k);
      return reversedHead;
    }
    return head;
  }
  ListNode? reverseLinkedList(ListNode? head, int k) {
    ListNode? prev = null;
    ListNode? curr = head;
    int remainingK = k;
    while (remainingK > 0 && curr != null) {
      final next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
      remainingK--;
    }
    return prev;
  }
}

File: sol/solution/0001-0100/0025/Solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseKGroup(head: ListNode?, k: Int): ListNode? {
        var curr = head
        var count = 0
        while (curr != null && count < k) {
            curr = curr.next
            count++
        }
        if (count == k) {
            val reversedHead = reverseLinkedList(head, k)
            // Recursively process the remaining list
            head?.next = reverseKGroup(curr, k)
            return reversedHead
        }
        return head
    }
    private fun reverseLinkedList(head: ListNode?, k: Int): ListNode? {
        var prev: ListNode? = null
        var curr = head
        var remainingK = k
        while (remainingK > 0 && curr != null) {
            val next = curr.next
            curr.next = prev
            prev = curr
            curr = next
            remainingK--
        }
        return prev
    }
}